hal/api/Emac.h

```diff
@@ -0,0 +1,173 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_EMAC_H
+#define MBED_EMAC_H
+
+#include "platform.h"
+
+#if DEVICE_EMAC
+
+#include "stack_mem.h"
+
+typedef struct emac_interface emac_interface_t;
+
+/**
+ * EmacInterface
+ *
+ * This interface should be used to abstract low level access to networking hardware
+ */
+
+/**
+ * Network interface capabilities
+ *
+ * Right now the flags are LWIP specific, but in the future we may need to support more stacks.
+ */
+#define EMAC_CAP_BROADCAST    0x01U     /*!< Interface has broadcast capabilities */
+#define EMAC_CAP_ETHARP       0x02U     /*!< Interface supports ARP */
+#define EMAC_CAP_ETHERNET     0x04U     /*!< Ethernet device */
+#define EMAC_CAP_IGMP         0x08U     /*!< Interface has IGMP capabilities */
+
+/**
+ * Callback to be register with Emac interface and to be called fore received packets
+ *
+ * @param data Arbitrary user data (IP stack)
+ * @param buf  Received data
+ */
+typedef void (*emac_link_input_fn)(void *data, stack_mem_t *buf);
```

Comment:
using ``stack_mem_t`` instead of ``stack_mem_chain_t`` would make things easier for the driver, since all the data would be in one contiguous block. The downside to this is zero copy wouldn't be fully supported, as the buffers would need to coalesced.

It seems reasonable that we could pin down at least a partial ``stack_mem_chain_t``. I have this mocked up for FNET and LWIP. The only functionality needed now is the ability to get a handle to each buffer in a memory chain, and the ability to free that buffer. The current API should provide this. This should make the code a bit more future proof, since no API change would be necissary for zero copy. This is really just an API optimization though, so I'm fine with it either way.